File: DSA/pranavdixit_590014418/Day_55_Question_1.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 10000

typedef struct {
    int key;
    int value;
} Pair;

int find(Pair* map, int size, int key) {
    for (int i = 0; i < size; i++)
        if (map[i].key == key)
            return map[i].value;
    return -1;
}

void insert(Pair* map, int* size, int key, int value) {
    if (find(map, *size, key) == -1) {
        map[*size].key = key;
        map[*size].value = value;
        (*size)++;
    }
}

int longestSubarraySumK(int arr[], int n, int k) {
    Pair map[MAX];
    int mapSize = 0;
    int sum = 0, maxLen = 0;

    for (int i = 0; i < n; i++) {
        sum += arr[i];

        if (sum == k)
            maxLen = i + 1;

        int prevIndex = find(map, mapSize, sum - k);
        if (prevIndex != -1)
            if (i - prevIndex > maxLen)
                maxLen = i - prevIndex;

        insert(map, &mapSize, sum, i);
    }

    return maxLen;
}

int main() {
    int arr[] = {10, 5, 2, 7, 1, 9};
    int k = 15;
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("Longest subarray length: %d\n", longestSubarraySumK(arr, n, k)); // Output: 4
    return 0;
}

File: DSA/upama_kumari_590014909/day_55_question_1.py
def longest_subarray_sum_k(arr, k):
    prefix_sum = 0
    prefix_index_map = {}
    max_len = 0

    for i, num in enumerate(arr):
        prefix_sum += num

        if prefix_sum == k:
            max_len = max(max_len, i + 1)

        if (prefix_sum - k) in prefix_index_map:
            max_len = max(max_len, i - prefix_index_map[prefix_sum - k])

        if prefix_sum not in prefix_index_map:
            prefix_index_map[prefix_sum] = i

    return max_len

print(longest_subarray_sum_k([10, 5, 2, 7, 1, 9], 15))  
print(longest_subarray_sum_k([1, 2, 3], 3)) 
